src/crop.sol

```diff
@@ -199,4 +218,11 @@ contract CropJoin {
 
         emit Tack(src, dst, wad);
     }
+
+    // msg.sender must vat.hope(adapter)
+    function frob(int256 dink, int256 dart) external {
+        address urp = proxy[msg.sender];
+        require(urp != address(0), "CropJoin/no-urn-proxy");
+        vat.frob(ilk, urp, urp, msg.sender, dink, dart);
```

Comment:
`msg.sender` could actually be relaxed to an argument as minting dai to an address is fine for arbitrary addresses, and `vat.hope(adapter)` is required to repay dai.

Reply:
We could actually maybe even open up all arguments provided we add checks that the proxy owner has "wish"ed `msg.sender`, but perhaps simpler is better to start.

Reply:
That's what I brought up in chat--it's not quite trivial, since if one just does the obvious thing and makes `w` an argument, then as soon as you `hope` the adapter, anyone can steal your DAI to repay their own CropJoin Vault (since they can just specify your address as `w`). The solution is to build some access control into the adapter itself (much like the CdpManager already does). So `CropJoin.frob` would check something like `require(dart > 0 || w == msg.sender || allowed(w, msg.sender))`.

I think it depends on the timeline we want to ship in. If we're trying to get this out by, say, the end of next week, I prefer to keep it as simple as possible given all the other moving pieces. If we're okay to let this slip a little more, e.g. to early July, then I like the idea of recovering more of the lost functionality.

Reply:
Ah yes you're right. Simpler is better for now because I would still like to hit the next week target if possible. We can come at a V2 of this with a more robust solution later.